src/future/timeout.rs

```diff
@@ -0,0 +1,78 @@
+use std::error::Error;
+use std::fmt;
+use std::pin::Pin;
+use std::time::Duration;
+
+use futures_timer::Delay;
+use pin_utils::unsafe_pinned;
+
+use crate::future::Future;
+use crate::task::{Context, Poll};
+
+/// Awaits a future or times out after a duration of time.
+///
+/// # Examples
+///
+/// ```no_run
+/// # #![feature(async_await)]
+/// # fn main() -> std::io::Result<()> { async_std::task::block_on(async {
+/// #
+/// use std::time::Duration;
+///
+/// use async_std::future;
+///
+/// let never = future::pending::<()>();
+/// let dur = Duration::from_secs(5);
+/// assert!(future::timeout(dur, never).await.is_err());
+/// #
+/// # Ok(()) }) }
+/// ```
+pub async fn timeout<F, T>(dur: Duration, f: F) -> Result<T, TimeoutError>
+where
+    F: Future<Output = T>,
+{
+    let f = TimeoutFuture {
+        future: f,
+        delay: Delay::new(dur),
+    };
+    f.await
+}
+
+/// A future that times out after a duration of time.
+#[doc(hidden)]
+#[allow(missing_debug_implementations)]
+struct TimeoutFuture<F> {
+    future: F,
+    delay: Delay,
+}
+
+impl<F> TimeoutFuture<F> {
+    unsafe_pinned!(future: F);
+    unsafe_pinned!(delay: Delay);
+}
+
+impl<F: Future> Future for TimeoutFuture<F> {
+    type Output = Result<F::Output, TimeoutError>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.as_mut().future().poll(cx) {
```

Comment:
Does it mean that in case of a CPU-bound future that is blocking the current thread you will never try to poll the timer and return TimeoutError?

Reply:
@kpp That's a good point - do you think we should check for timeout before polling the future?

I worry a bit that this would add overhead to futures that are quick to poll or immediately ready. And if the future is CPU-bound and takes a really long time, we can still get stuck if the timeout occurs while that future is running.

All things considered, I think I prefer the current behavior. Wdyt?

Reply:
> All things considered, I think I prefer the current behavior.

Yep... I don't see a better solution for this.

Still I would rewrite the current implementation as `f.select(Delay::new(dur))...` ;)